### PR TITLE
Merge develop back into master

### LINKED ISSUE #37: `AttackerModel` doesn't do normal inference
Hi there,
Great work on this package! Forgive me if I'm missing something, but there seems to be a mismatch between how `AttackerModel.forward` works, and the documentation. I want to call `AttackerModel.forward` and just do normal inference without returning any image. I should be able to do this by setting `make_adv=False` and `with_image=False`. But as I can see [here](https://github.com/MadryLab/robustness/blob/6347646ae47120d35f47220eabd507cebbd6c914/robustness/attacker.py#L322) this just sets the output to `None`! Moreover, it still outputs a tuple with the input image as the second element. 

Finally, the docstring contains a broken link: "see `our detailed
walkthrough <../example_usage/input_space_manipulation>`".

As an addendum, you might consider making the default behavior being to just do normal inference and not attack or output an image. This may make it easier to incorporate an `AttackerModel` object as a step in other modules.
Thanks!
-N

### LINKED ISSUE #41: Issue with resuming training from a checkpoint
I noticed that when resuming training from checkpoint using the CLI, the checkpoint weights are loaded, but the current epoch and learning schedule are not resumed. 

Unless I'm missing something, it looks like it's because the checkpoint is not passed to `train_model`; see https://github.com/MadryLab/robustness/blob/6347646ae47120d35f47220eabd507cebbd6c914/robustness/main.py#L57

Seems like an easy fix, as the infrastructure to receive the checkpoint in `train_model` exists -- when I pass the checkpoint myself, I only get one other error, from https://github.com/MadryLab/robustness/blob/6347646ae47120d35f47220eabd507cebbd6c914/robustness/train.py#L258 where it looks like the checkpoint doesn't have natural or adversarial scores saved.

Thoughts?

### PR DESCRIPTION
**Pull request by andrewilyas**
Probably enough for version 1.2 of the library?

- Fixed `random_perturb` for robust models
- Add `random_step` (useful for training randomized smoothing models)
- *Mixed precision support* but without resuming
- Fix #37 (`with_image` option in AttackerModel)
- Allow for epsilon schedules (removed the `eps-fadein` argument in favor of a more general epsilon multiplier schedule)
- Allow resuming optimizer instead of just model with `--resume-optimizer` (Fix #41)
- `make_and_restore_model` now makes the model parallel even if resume_path is not specified, new `add_custom_forward` option for better interfacing with custom architectures
- New dataset (Places365)
- Improved learning rate scheduling

### CODE REVIEW

**Comment by andrewilyas** [comment]
Ok so right off the bat a couple of things we should fix before we merge:

- Right now parallel=True is on by default, but train expects a model that is *not* DataParallel, we should fix this (one solution is to make the default value of parallel to be `(resume_path is not None)`)
- We should consider evaluating the model straightaway upon resume to get good values of prec1 that are unaffected by the first epoch of training
- We should make sure all of the new additions are sufficiently documented (specifically a short addition to the training doc RE mixed precision would probably be useful)

**Comment by andrewilyas** [comment]
Ok the only immediately missing things before I can merge this in:

- [ ] I'm not fully convinced of the ``best_prec1`` thing---could see some problems arising if someone is training adversarially and then wants to resume but with standard training. I think it should still be named either ``adv_prec1`` or ``nat_prec1`` (or something like that, and then we should just have a check for whether it's in the checkpoint).

- [ ] Need to update the docs for CLI training (everything else should be basically documented)
